source/stun_deserializer.c

```diff
@@ -0,0 +1,868 @@
+    if( result == STUN_RESULT_OK )
+    {
+        class = pAttribute->pAttributeValue[STUN_ERROR_CODE_PACKET_ERROR_CLASS_OFFSET];
+        errorNumber = pAttribute->pAttributeValue[STUN_ERROR_CODE_PACKET_ERROR_CODE_OFFSET];
+
+        *errorCode = STUN_GET_ERROR( class,
+                                     errorNumber );
+        *errorPhrase = ( uint8_t * ) &( pAttribute->pAttributeValue[STUN_ERROR_CODE_PACKET_ERROR_PHRASE_OFFSET] );
+    }
+    return result;
+}
+
+/*-----------------------------------------------------------*/
+
+StunResult_t StunDeserializer_ParseAttributeChannelNumber( const StunContext_t * pCtx,
+                                                           const StunAttribute_t * pAttribute,
+                                                           uint16_t * channelNumber,
+                                                           StunAttributeType_t attributeType )
+{
+    StunResult_t result = STUN_RESULT_OK;
+
+    if( ( pAttribute == NULL ) ||
+        ( channelNumber == NULL ) ||
+        ( pAttribute->attributeType != attributeType ) ||
+        ( pAttribute->pAttributeValue == NULL ) )
+    {
+        result = STUN_RESULT_BAD_PARAM;
+    }
+
+    if( result == STUN_RESULT_OK )
+    {
+        if( pAttribute->attributeValueLength != STUN_ATTRIBUTE_CHANNEL_NUMBER_LENGTH )
+        {
+            result = STUN_RESULT_INVALID_ATTRIBUTE_LENGTH;
+        }
+    }
+
+    if( result == STUN_RESULT_OK )
+    {
+        STUN_READ_UINT16( channelNumber, ( uint8_t * ) &( *( ( uint16_t * ) pAttribute->pAttributeValue ) ) );
+    }
+
+    return result;
+}
+/*-----------------------------------------------------------*/
+
+StunResult_t StunDeserializer_ParseAttributeUseCandidate( StunContext_t * pCtx,
+                                                          const StunAttribute_t * pAttribute,
+                                                          StunAttributeType_t attributeType )
+{
+    StunResult_t result = STUN_RESULT_OK;
+
+    if( ( pAttribute == NULL ) ||
+        ( pAttribute->attributeType != attributeType ) ||
+        ( pAttribute->pAttributeValue != NULL ) )
+    {
+        result = STUN_RESULT_BAD_PARAM;
+    }
+
+    if( result == STUN_RESULT_OK )
+    {
+        if( pAttribute->attributeValueLength != 0 )
+        {
+            result = STUN_RESULT_INVALID_ATTRIBUTE_LENGTH;
+        }
+    }
+
+    if( result == STUN_RESULT_OK )
+    {
+        pCtx->attributeFlag |= STUN_FLAG_USE_CANDIDATE_ATTRIBUTE;
+    }
+
+    return result;
+}
+/*-----------------------------------------------------------*/
+
+StunResult_t StunDeserializer_ParseAttributeDontFragment( StunContext_t * pCtx,
+                                                          const StunAttribute_t * pAttribute,
+                                                          StunAttributeType_t attributeType )
+{
+    StunResult_t result = STUN_RESULT_OK;
+
+    if( ( pAttribute == NULL ) ||
+        ( pAttribute->attributeType != attributeType ) ||
```

Comment:
Does the application confirm, StunContext passed is NOT NULL?

Reply:
Added NULL check, thanks